ta/fde_key_handler/fde_key_handler_ta_handle.c

```diff
@@ -0,0 +1,297 @@
+/* SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2023, Canonical Ltd.
+ */
+
+#include <assert.h>
+#include <util.h>
+#include <tee_internal_api.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+#include <tee_internal_api_extensions.h>
+#include <pta_system.h>
+#include <string.h>
+#include <string_ext.h>
+#include <trace.h>
+
+#include "fde_key_handler_ta_type.h"
+#include "fde_key_handler_ta_handle.h"
+
+#define IV_SIZE         16
+#define NONCE_SIZE      32
+#define TAG_SIZE        16
+#define KEY_HANDLE_VERSION 'U'
+
+// Trusted Key Handle
+struct key_handle {
+    uint8_t version; // reserved to recognise the version
+    uint8_t iv[IV_SIZE];
+    uint8_t nonce[NONCE_SIZE];
+    uint8_t tag[TAG_SIZE];
+};
+
+/**
+ * Derive TA unique key
+ * use passed handle as extra data for key derivation
+ *  key: buffer to created key
+ *  key_size: size of derived key buffer
+ *  handle: buffer with handle
+ *  handle_size: size handle buffer
+ */
+static TEE_Result derive_ta_unique_key(uint8_t *key,
+                                       uint16_t key_size,
+                                       uint8_t *handle,
+                                       uint32_t handle_size) {
+
+  TEE_TASessionHandle sess = TEE_HANDLE_NULL;
+  TEE_Param params[TEE_NUM_PARAMS];
+  TEE_Result res = TEE_ERROR_GENERIC;
+  uint32_t ret_orig = 0;
+  uint32_t param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+                                         TEE_PARAM_TYPE_MEMREF_OUTPUT,
+                                         TEE_PARAM_TYPE_NONE,
+                                         TEE_PARAM_TYPE_NONE);
+
+  res = TEE_OpenTASession(&(const TEE_UUID)PTA_SYSTEM_UUID,
+                          TEE_TIMEOUT_INFINITE,
+                          0,
+                          NULL,
+                          &sess,
+                          &ret_orig);
+  if (res) {
+    return res;
+  }
+
+  if (handle && handle_size) {
+    params[0].memref.buffer = handle;
+    params[0].memref.size = handle_size;
+  }
+
+  params[1].memref.buffer = key;
+  params[1].memref.size = key_size;
+
+  res = TEE_InvokeTACommand(sess,
+                            TEE_TIMEOUT_INFINITE,
+                            PTA_SYSTEM_DERIVE_TA_UNIQUE_KEY,
+                            param_types,
+                            params,
+                            &ret_orig);
```

Comment:
Why is this called all the time? Should not it be just called once, on initialization?

Reply:
got it, thanks